### PR TITLE
[SPARK-49241][CORE] Add `OpenTelemetryPush` Sink with `opentelemetry` profile

### PR DESCRIPTION
**Pull request by TQJADE**
### What changes were proposed in this pull request?
OpenTelemetryPushSink and OpenTelemetryPushReporter has been added in the supported Sink Classes.

### Why are the changes needed?

- From OpenTelemetry side, they have added the support from a pull model: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/apachesparkreceiver

- However, for a very short jobs, only the push model is applicable. Therefore, a new sink OpenTelemetryPushSink and a new reporter OpenTelemetryPushReporter have been introduced here.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
* Build the Apache Spark Core Snapshot. Reference the Snapshot and giving the local OpenTelemetry endpoint `127.0.0.1:4317` as the endpoint for OpenTelemetryReporter.
* Start the open-telemetry collector locally, per: https://opentelemetry.io/docs/collector/quick-start/#generate-and-collect-telemetry

```
docker run \
  -p 127.0.0.1:4317:4317 \
  -p 127.0.0.1:55679:55679 \
  otel/opentelemetry-collector:0.107.0 \
  2>&1 | tee collector-output.txt # Optionally tee output for easier search later
```
  
* Start Spark Applications and publishing the metrics
* Observing the metrics to be printed. 

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Thank you for making a PR, @TQJADE .

**Review by dongjoon-hyun — COMMENTED** [review]
Could you make CIs happy?

**Comment by dongjoon-hyun** [comment]
It seems that there are multiple failures in CIs. Do you think those failures related to this PR?

File: core/pom.xml [hunk 1]
```diff
@@ -616,6 +616,26 @@
         <script.extension>.sh</script.extension>
       </properties>
     </profile>
+    <profile>
+      <id>opentelemetry-reporter</id>
+      <dependencies>
+        <dependency>
+          <groupId>io.opentelemetry</groupId>
+          <artifactId>opentelemetry-sdk</artifactId>
+          <version>1.40.0</version>
```
Comment (dongjoon-hyun) [inline core/pom.xml:625]: Let's use the latest one.
- https://mvnrepository.com/artifact/io.opentelemetry/opentelemetry-api/1.41.0

Comment (TQJADE) [reply core/pom.xml]: Updated

**Comment by TQJADE** [comment]
> 1. It seems that there are multiple failures in CIs. Do you think those failures related to this PR?
> 2. When we enable this profile, it seems that this PR introduces many dependencies. Do we need all?
> 
> ```
> $ dev/make-distribution.sh -Popentelemetry-reporter
> $ ls -alh dist/jars/*opentelemetry*
> -rw-r--r--@ 1 dongjoon  staff   138K Aug 16 09:20 dist/jars/opentelemetry-api-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    42K Aug 16 09:20 dist/jars/opentelemetry-api-incubator-1.40.0-alpha.jar
> -rw-r--r--@ 1 dongjoon  staff    46K Aug 16 09:20 dist/jars/opentelemetry-context-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   100K Aug 16 09:20 dist/jars/opentelemetry-exporter-common-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    68K Aug 16 09:20 dist/jars/opentelemetry-exporter-otlp-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   188K Aug 16 09:20 dist/jars/opentelemetry-exporter-otlp-common-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    25K Aug 16 09:20 dist/jars/opentelemetry-exporter-sender-okhttp-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   6.6K Aug 16 09:20 dist/jars/opentelemetry-sdk-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    53K Aug 16 09:20 dist/jars/opentelemetry-sdk-common-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    19K Aug 16 09:20 dist/jars/opentelemetry-sdk-extension-autoconfigure-spi-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    53K Aug 16 09:20 dist/jars/opentelemetry-sdk-logs-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   302K Aug 16 09:20 dist/jars/opentelemetry-sdk-metrics-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   125K Aug 16 09:20 dist/jars/opentelemetry-sdk-trace-1.40.0.jar
> ```
> 
>  3. Does this introduce any transitive dependencies too?



> 1. It seems that there are multiple failures in CIs. Do you think those failures related to this PR?
> 2. When we enable this profile, it seems that this PR introduces many dependencies. Do we need all?
> 
> ```
> $ dev/make-distribution.sh -Popentelemetry-reporter
> $ ls -alh dist/jars/*opentelemetry*
> -rw-r--r--@ 1 dongjoon  staff   138K Aug 16 09:20 dist/jars/opentelemetry-api-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    42K Aug 16 09:20 dist/jars/opentelemetry-api-incubator-1.40.0-alpha.jar
> -rw-r--r--@ 1 dongjoon  staff    46K Aug 16 09:20 dist/jars/opentelemetry-context-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   100K Aug 16 09:20 dist/jars/opentelemetry-exporter-common-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    68K Aug 16 09:20 dist/jars/opentelemetry-exporter-otlp-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   188K Aug 16 09:20 dist/jars/opentelemetry-exporter-otlp-common-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    25K Aug 16 09:20 dist/jars/opentelemetry-exporter-sender-okhttp-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   6.6K Aug 16 09:20 dist/jars/opentelemetry-sdk-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    53K Aug 16 09:20 dist/jars/opentelemetry-sdk-common-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    19K Aug 16 09:20 dist/jars/opentelemetry-sdk-extension-autoconfigure-spi-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    53K Aug 16 09:20 dist/jars/opentelemetry-sdk-logs-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   302K Aug 16 09:20 dist/jars/opentelemetry-sdk-metrics-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   125K Aug 16 09:20 dist/jars/opentelemetry-sdk-trace-1.40.0.jar
> ```
> 
>  3. Does this introduce any transitive dependencies too?



> 1. It seems that there are multiple failures in CIs. Do you think those failures related to this PR?
> 2. When we enable this profile, it seems that this PR introduces many dependencies. Do we need all?
> 
> ```
> $ dev/make-distribution.sh -Popentelemetry-reporter
> $ ls -alh dist/jars/*opentelemetry*
> -rw-r--r--@ 1 dongjoon  staff   138K Aug 16 09:20 dist/jars/opentelemetry-api-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    42K Aug 16 09:20 dist/jars/opentelemetry-api-incubator-1.40.0-alpha.jar
> -rw-r--r--@ 1 dongjoon  staff    46K Aug 16 09:20 dist/jars/opentelemetry-context-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   100K Aug 16 09:20 dist/jars/opentelemetry-exporter-common-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    68K Aug 16 09:20 dist/jars/opentelemetry-exporter-otlp-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   188K Aug 16 09:20 dist/jars/opentelemetry-exporter-otlp-common-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    25K Aug 16 09:20 dist/jars/opentelemetry-exporter-sender-okhttp-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   6.6K Aug 16 09:20 dist/jars/opentelemetry-sdk-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    53K Aug 16 09:20 dist/jars/opentelemetry-sdk-common-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    19K Aug 16 09:20 dist/jars/opentelemetry-sdk-extension-autoconfigure-spi-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff    53K Aug 16 09:20 dist/jars/opentelemetry-sdk-logs-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   302K Aug 16 09:20 dist/jars/opentelemetry-sdk-metrics-1.40.0.jar
> -rw-r--r--@ 1 dongjoon  staff   125K Aug 16 09:20 dist/jars/opentelemetry-sdk-trace-1.40.0.jar
> ```
> 
> 3. Does this introduce any transitive dependencies too?

Reduced to below
```
 apache-spark git:(sink) ls -alh dist/jars/*opentelemetry*                                                               git:(sink↑14|) 
-rw-r--r--  1 qitan  staff   139K Aug 22 14:10 dist/jars/opentelemetry-api-1.41.0.jar
-rw-r--r--  1 qitan  staff    46K Aug 22 14:10 dist/jars/opentelemetry-context-1.41.0.jar
-rw-r--r--  1 qitan  staff   101K Aug 22 14:10 dist/jars/opentelemetry-exporter-common-1.41.0.jar
-rw-r--r--  1 qitan  staff    77K Aug 22 14:10 dist/jars/opentelemetry-exporter-otlp-1.41.0.jar
-rw-r--r--  1 qitan  staff   201K Aug 22 14:10 dist/jars/opentelemetry-exporter-otlp-common-1.41.0.jar
-rw-r--r--  1 qitan  staff   6.6K Aug 22 14:10 dist/jars/opentelemetry-sdk-1.41.0.jar
-rw-r--r--  1 qitan  staff    53K Aug 22 14:10 dist/jars/opentelemetry-sdk-common-1.41.0.jar
-rw-r--r--  1 qitan  staff   303K Aug 22 14:10 dist/jars/opentelemetry-sdk-metrics-1.41.0.jar
```

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for updating the PR.

File: core/pom.xml [hunk 2]
```diff
@@ -616,6 +626,52 @@
         <script.extension>.sh</script.extension>
       </properties>
     </profile>
+    <profile>
+      <id>opentelemetry-reporter</id>
```
Comment (dongjoon-hyun) [inline core/pom.xml:630]: It seems too long. Shall we use `opentelemetry` instead of `opentelemetry-reporter`?

File: project/SparkBuild.scala [hunk 3]
```diff
@@ -1329,6 +1333,13 @@ object Volcano {
   )
 }
 
+object OpenTelemetry {
+  // Exclude all OpenTelemetry file for Compile and Test
```
Comment (dongjoon-hyun) [inline project/SparkBuild.scala:1337]: Although it comes from `object Volcano`, let's use `files` instead of `file`.

File: pom.xml [hunk 4]
```diff
@@ -3065,6 +3066,9 @@
               <javacArg>${java.version}</javacArg>
               <javacArg>-Xlint:all,-serial,-path,-try</javacArg>
             </javacArgs>
+            <excludes>
+              <exclude>${opentelemetry.exclude}</exclude>
+            </excludes>
```
Comment (dongjoon-hyun) [inline pom.xml:3071]: Shall we move this `core/pom.xml` instead of here?

File: pom.xml [hunk 5]
```diff
@@ -331,6 +331,7 @@
     <ojdbc11.version>23.5.0.24.07</ojdbc11.version>
     <!-- Used for SBT build to retrieve the Spark version -->
     <spark.version>${project.version}</spark.version>
+    <opentelemetry.exclude>**/OpenTelemetry*.scala</opentelemetry.exclude>
```
Comment (dongjoon-hyun) [inline pom.xml:334]: Shall we move this to `core/pom.xml`?

File: core/src/test/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporterSuite.scala [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+package org.apache.spark.metrics.sink.opentelemetry
+
+import com.codahale.metrics.MetricRegistry
+import org.scalatest.PrivateMethodTester
+
+import org.apache.spark.SparkFunSuite
+
+class OpenTelemetryPushReporterSuite
+    extends SparkFunSuite
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporterSuite.scala:26]: We need 2-space indentation here.
- https://github.com/databricks/scala-style-guide

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+    pollInterval: Int = 10,
+    pollUnit: TimeUnit = TimeUnit.SECONDS,
+    host: String = "http://localhost",
+    port: String = "4317",
+    headersMap: Map[String, String] = Map(),
+    attributesMap: Map[String, String] = Map(),
+    trustedCertificatesPath: String,
+    privateKeyPemPath: String,
+    certificatePemPath: String
+) extends ScheduledReporter (
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:47]: Please move the starting `)` to the end of the above line.
- https://github.com/databricks/scala-style-guide

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+      registry,
+      "opentelemetry-push-reporter",
+      MetricFilter.ALL,
+      TimeUnit.SECONDS,
+      TimeUnit.MILLISECONDS
+    )
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:53]: ditto.

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+
+  val sdkMeterProvider: SdkMeterProvider = SdkMeterProvider
+    .builder()
+    .registerMetricReader(metricReader)
+    .setResource(resource)
+    .build()
+
+  val openTelemetryCounters = collection.mutable.Map[String, LongCounter]()
+  val openTelemetryHistograms =
+    collection.mutable.Map[String, DoubleHistogram]()
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:121]: May I ask why this statement has two lines?

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+  for ((key, value) <- attributesMap) {
+    arrtributesBuilder.put(key, value)
+  }
+
+  val resource = Resource
+    .getDefault()
+    .merge(Resource.create(arrtributesBuilder.build()));
+
+  val metricReader =
+    PeriodicMetricReader
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:108]: Please move `PeriodicMetricReader` to the above line.

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+    .buildAndRegisterGlobal();
+  val openTelemetryMeter = openTelemetry.getMeter("apache-spark")
+
+  override def report(
+      gauges: SortedMap[String, Gauge[_]],
+      counters: SortedMap[String, Counter],
+      histograms: SortedMap[String, Histogram],
+      meters: SortedMap[String, Meter],
+      timers: SortedMap[String, Timer]
+  ): Unit = {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:139]: ditto.

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushSink.scala [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+import com.codahale.metrics.MetricRegistry
+import org.apache.commons.lang3.StringUtils
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.metrics.sink.Sink
+
+
+private[spark] object OpenTelemetryPushSink {
+  private def fetchMapFromProperties(
+                                      properties: Properties,
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushSink.scala:32]: Could you double-check the coding style?

File: build/sbt [hunk 13]
```diff
@@ -94,7 +99,7 @@ process_my_args () {
          -batch) exec </dev/null && shift ;;
 
     -sbt-create) sbt_create=true && shift ;;
-
+    -P*) enableProfile "$1" && shift ;;
```
Comment (dongjoon-hyun) [inline build/sbt:102]: Do we still need this?

This is an orthogonal contribution. If we still need this, we should make an independent PR, @TQJADE .

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+    pollInterval: Int = 10,
+    pollUnit: TimeUnit = TimeUnit.SECONDS,
+    host: String = "http://localhost",
+    port: String = "4317",
+    headersMap: Map[String, String] = Map(),
+    attributesMap: Map[String, String] = Map(),
+    trustedCertificatesPath: String,
+    privateKeyPemPath: String,
+    certificatePemPath: String)
+    extends ScheduledReporter(
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:46]: 2-space indentation.

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+      registry,
+      "opentelemetry-push-reporter",
+      MetricFilter.ALL,
+      TimeUnit.SECONDS,
+      TimeUnit.MILLISECONDS)
+    with MetricRegistryListener {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:52]: ditto.

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushSink.scala [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+        val mapKey = key.substring(dotIndex + 1)
+        propertiesMap(mapKey) = properties.getProperty(key)
+      }
+    }
+    propertiesMap.toMap
+  }
+}
+
+private[spark] class OpenTelemetryPushSink(val property: Properties, val registry: MetricRegistry)
+    extends Sink
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushSink.scala:49]: 2-space indentation.

Comment (TQJADE) [reply core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushSink.scala:49]: after running `./dev/scalafmt`, the file is auto recorrected to four spaces.

Comment (TQJADE) [reply core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushSink.scala]: Changed in https://github.com/apache/spark/pull/47763/commits/b4ae2df0d9bd4d75df54a53f596d6b6045c8c354

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala (hunk 15 above)
Comment (TQJADE) [reply core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala]: Changed in https://github.com/apache/spark/pull/47763/commits/b4ae2df0d9bd4d75df54a53f596d6b6045c8c354

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala (hunk 14 above)
Comment (TQJADE) [reply core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala]: changed in https://github.com/apache/spark/pull/47763/commits/b4ae2df0d9bd4d75df54a53f596d6b6045c8c354

File: build/sbt (hunk 13 above)
Comment (TQJADE) [reply build/sbt]: removed

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]

File: core/src/test/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporterSuite.scala [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package org.apache.spark.metrics.sink.opentelemetry
+
+import com.codahale.metrics.MetricRegistry
+import org.scalatest.PrivateMethodTester
+
+import org.apache.spark.SparkFunSuite
+
+class OpenTelemetryPushReporterSuite extends SparkFunSuite with PrivateMethodTester {
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporterSuite.scala:25]: Could you add more test coverage?

File: core/pom.xml [hunk 18]
```diff
@@ -616,6 +626,36 @@
...
+        </dependency>
+        <dependency>
+          <groupId>io.opentelemetry</groupId>
+          <artifactId>opentelemetry-sdk</artifactId>
+          <version>1.41.0</version>
+        </dependency>
+        <dependency>
+          <groupId>com.squareup.okhttp3</groupId>
+          <artifactId>okhttp</artifactId>
+          <version>3.13.0</version>
```
Comment (dongjoon-hyun) [inline core/pom.xml:654]: Please note that Apache Spark's `okhttp` is 3.12.12. So, we had better use 3.12.12 even for the test case.

https://github.com/apache/spark/blob/9fc1e05224d2b4207885a1d4f822fed6db0cb1a1/dev/deps/spark-deps-hadoop-3-hive-2.3#L226

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Thank you for updating this PR. We still need your replies for the following review comments.
- https://github.com/apache/spark/pull/47763#discussion_r1730446753
- https://github.com/apache/spark/pull/47763#discussion_r1732172773

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala [hunk 19]
```diff
@@ -0,0 +1,360 @@
...
+import io.opentelemetry.sdk.metrics.SdkMeterProvider
+import io.opentelemetry.sdk.metrics.export.PeriodicMetricReader
+import io.opentelemetry.sdk.resources.Resource
+
+private[spark] class OpenTelemetryPushReporter(
+    registry: MetricRegistry,
+    pollInterval: Int = 10,
+    pollUnit: TimeUnit = TimeUnit.SECONDS,
+    host: String = "http://localhost",
+    port: String = "4317",
```
[the remark this answers is not in the document]
Comment (TQJADE) [reply core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:38]: @dongjoon-hyun 4317 is port for grpc and 4318 is for HTTP/HTTPS. Should I rename the port to `grpcPort`?

https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/

[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:38]: To @TQJADE , you are using `http://` in line 39, aren't you? Are you using `grpc` instead of HTTP here?

File: core/pom.xml [hunk 20]
```diff
@@ -651,7 +651,8 @@
         <dependency>
           <groupId>com.squareup.okhttp3</groupId>
           <artifactId>okhttp</artifactId>
-          <version>3.13.0</version>
+          <!--version should be same with https://github.com/apache/spark/blob/0895471ad3b2a82668ea7e83dd20dda6b17f145f/dev/deps/spark-deps-hadoop-3-hive-2.3#L226-->
```
Comment (dongjoon-hyun) [inline core/pom.xml:654]: Please remove this comment.

Comment (TQJADE) [reply core/pom.xml]: Removed

File: core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala (hunk 19 above)
[the remark this answers is not in the document]
Comment (TQJADE) [reply core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:38]: https://github.com/open-telemetry/opentelemetry-java/blob/7522bfe141524d52649b5c344c643104b57af608/exporters/common/src/main/java/io/opentelemetry/exporter/internal/ExporterBuilderUtil.java#L32-L47
Here there is a endpoint validation, it required the endpoint to start with http/https

[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:38]: What does that mean? Do you mean `OpenTelemetry` requires `HTTP` with 4317 ? If we need to use `4317` with HTTP, which HTTP call use 4318 port?

[the remark this answers is not in the document]
Comment (TQJADE) [reply core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:38]: Here I am using grpc endpoint, when I tested, the endpoint should be started with `http` or `https`. The default value of Opentelmetry GRPC endpoint port is 4317. HTTP endpoint port is 4318.

[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/metrics/sink/opentelemetry/OpenTelemetryPushReporter.scala:38]: It seems that we are repeating the same questions and answer again and again. :) 

What I asked to confirm is that why we are not using the default value of HTTP endpoint port (4318) here when the endpoint starts with `http`. This PR uses `http://localhost:4317` instead of `http://localhost:4318`, doesn't it? We know that your sentence is correct, but it sounds like contradiction. We need an explanation for this contradiction.
>  the endpoint should be started with http or https. The default value of Opentelmetry GRPC endpoint port is 4317. HTTP endpoint port is 4318.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @TQJADE .
Merged to master.

**Comment by LuciferYang** [comment]
Is it necessary to activate the `opentelemetry` profile to execute unit tests in IntelliJ, regardless of whether it's Maven or SBT? I've found that without activating the `opentelemetry` profile, I'm unable to execute unit tests in IntelliJ. For example, when I try to execute `DataSourceV2SQLSuiteV2Filter` in IntelliJ, I encounter the following error:

image
. 
Although manually activating the `opentelemetry` profile can serve as a workaround, this is not a good experience because `opentelemetry` is not activated by default. Or are there any other correct ways to ensure it works?

cc @TQJADE @dongjoon-hyun 

also cc @panbingkun

**Comment by dongjoon-hyun** [comment]
Got it. 

Actually, this PR follows the existing `Volcano` profile way which has the same side-effect. However, this seems more disruptive because this is in `core`.

To @TQJADE , could you find a better way to mitigate the above situation?

**Comment by mridulm** [comment]
QQ: Cant this not live outside of Apache Spark - and use the existing integrations in MetricSystem ?
I am trying to understand why we need to include it as part of Spark, and not as an external package - are the current extension mechanisms not sufficient ?

+CC @dongjoon-hyun

**Comment by LuciferYang** [comment]
> Got it.
> 
> Actually, this PR follows the existing `Volcano` profile way which has the same side-effect. However, this seems more disruptive because this is in `core` while `Volcano` affects only K8s module.
> 
> To @TQJADE , could you find a better way to mitigate the above situation?


cc @dongjoon-hyun @TQJADE In fact, due to our use of `sbt-pom-reader`, on the Maven side, we can utilize the `build-helper-maven-plugin` to dynamically load additional `src` and `test` code paths. The configuration for `build-helper-maven-plugin` is also dynamically activated in `SparkBuild.scala` without requiring any additional configuration. I think this method should resolve the issue I mentioned. 

I have submitted a pr: https://github.com/apache/spark/pull/47986. 

Before supplementing the PR description, I will take some time to conduct additional verification.


> I am trying to understand why we need to include it as part of Spark, and not as an external package - are the current extension mechanisms not sufficient ?

Of course, this question also requires a clear answer :)

**Comment by TQJADE** [comment]
> I am trying to understand why we need to include it as part of Spark, and not as an external package - are the current extension mechanisms not sufficient ?

@mridulm @LuciferYang 
We have an internal service which accepts the OpenTelemetrySDK to publish metrics. However, this service only allow push model, rather than the pull model to collect metrics. 
Meanwhile, we have tons of Spark Jobs running in K8s and they can integrate with the classes within org.apache.spark.metrics.sink.* to publish metrics easily. I found an existing push model sink class:
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/metrics/sink/StatsdSink.scala. It is used to "translate" the codahale metrics to statsd metrics so I draft this PR for the OpenTelemetry usage.

**Comment by TQJADE** [comment]
> I am trying to understand why we need to include it as part of Spark, and not as an external package - are the current extension mechanisms not sufficient ?

@mridulm @LuciferYang 
I would like to use OpenTelemetry instrumentations to “translate” the codahale metrics of Spark Jobs to opentelemetry metrics. However this is disabled by default as below with reasons:
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/supported-libraries.md#disabled-instrumentations

> Some instrumentations can produce too many spans and make traces very noisy. For this reason, the following instrumentations are disabled by default. dropwizard-metrics which might create a very low quality metrics data, because of lack of label/attribute support in the Dropwizard metrics API.

That is the reason I would like to contribute to Apache Spark side. The OpenTelemetryPushSink is similar to StatsdSink, to push the metrics to target services.
cc: @dongjoon-hyun

**Comment by mridulm** [comment]
I am trying to understand whether there are any limitations due to which this integration has to be included as part of Apache Spark.
In other words, can this be an external integration/library that users include for the spark apps which need it.

**Comment by dongjoon-hyun** [comment]
Your point is correct, @mridulm .
- There is no limitation in Spark architecture and this can be an external library without any problem.
- Initially, I misunderstood the context of the existing `opentelemetry`'s `Disabled instrumentations` as discontinued or deprecated ones. It's my bad.

Let me revert this commit. Thank you for all your valuable reviews, @mridulm , @LuciferYang . Personally, sorry for wasting your time, @TQJADE .

**Comment by dongjoon-hyun** [comment]
This is reverted via https://github.com/apache/spark/commit/628cd751a909e63a9df0bca317b6e611a736ff85

**Comment by mridulm** [comment]
Thanks for the details @dongjoon-hyun ! Really appreciate it.
